irohad/ametsuchi/impl/postgres_block_storage.cpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "ametsuchi/impl/postgres_block_storage.hpp"
+
+#include "common/hexutils.hpp"
+#include "logger/logger.hpp"
+
+using namespace iroha::ametsuchi;
+
+PostgresBlockStorage::PostgresBlockStorage(
+    soci::session &sql,
+    std::shared_ptr<BlockTransportFactory> block_factory,
+    logger::LoggerPtr log)
+    : sql_(sql),
+      block_factory_(std::move(block_factory)),
+      log_(std::move(log)) {}
+
+bool PostgresBlockStorage::insert(
+    std::shared_ptr<const shared_model::interface::Block> block) {
+  soci::statement st =
+      (sql_.prepare
+           << "INSERT INTO blocks(id, block_data) VALUES(:id, :block_data)",
+       soci::use(block->height()),
+       soci::use(block->blob().hex()));
+  log_->debug("insert: {}", block->blob().hex());
+  try {
+    st.execute(true);
+    return true;
+  } catch (const std::exception &e) {
+    log_->warn(
+        "Failed to insert block {}, reason {}", block->height(), e.what());
+    return false;
+  }
+}
+
+boost::optional<std::shared_ptr<const shared_model::interface::Block>>
+PostgresBlockStorage::fetch(
+    shared_model::interface::types::HeightType height) const {
+  using T = boost::tuple<std::string>;
+  auto result = execute<T>([&] {
+    return (sql_.prepare << "SELECT block_data FROM blocks WHERE id = :id",
+            soci::use(height));
+  });
+  return flatMapValue<
+      boost::optional<std::shared_ptr<const shared_model::interface::Block>>>(
+      result, [&](auto &block_data) {
+        log_->debug("fetched: {}", block_data);
+        auto byte_block = iroha::hexstringToBytestring(block_data);
+        if (not byte_block) {
+          return boost::optional<
+              std::shared_ptr<const shared_model::interface::Block>>(
+              boost::none);
+        }
+
+        iroha::protocol::Block_v1 block;
+        block.ParseFromString(*byte_block);
+        return block_factory_->build(std::move(block))
+            .match(
+                [&](auto &&v) {
+                  return boost::make_optional(
+                      std::shared_ptr<const shared_model::interface::Block>(
+                          std::move(v.value)));
+                },
+                [&](const auto &e)
+                    -> boost::optional<
+                        std::shared_ptr<const shared_model::interface::Block>> {
+                  log_->error("{}", e.error.error);
+                  return boost::none;
+                });
+      });
+}
+
+size_t PostgresBlockStorage::size() const {
+  using T = boost::tuple<shared_model::interface::types::HeightType>;
+  auto result = execute<T>(
+      [&] { return (sql_.prepare << "SELECT COUNT(*) FROM blocks"); });
+  return flatMapValue<
+             boost::optional<shared_model::interface::types::HeightType>>(
+             result, [](auto &count) { return boost::make_optional(count); })
+      .value_or(0);
```

Comment:
also, if we decide not to drop the table in `clear()`, unset `optional` should be logged here as an error, i think.

Reply:
I will be logged inside `execute` as error, I think that's enough.